### PR TITLE
Handle instance deletes as a queue job.

### PR DESCRIPTION
**Pull request by mikalstill**

### CODE REVIEW

File: shakenfist/daemons/triggers.py [hunk 1]
```diff
@@ -65,6 +65,7 @@ def run(self):
             all_observers = list(observers.keys())
             for instance_uuid in all_observers:
                 if not observers[instance_uuid].is_alive():
+                    observers[instance_uuid].join(1)
```
Comment (mandoonandy) [inline shakenfist/daemons/triggers.py:68]: Maybe a quick comment why join() a not is_alive() process. (Unless everybody knows but me... does happen)

File: shakenfist/external_api/app.py [hunk 2]
```diff
@@ -596,16 +562,27 @@ def get(self, instance_uuid=None, instance_from_db=None):
...
+        # If this instance is not on a node, just do the DB cleanup locally
+        if not instance_from_db['node']:
+            node = config.parsed.get('NODE_NAME')
+        else:
+            node = instance_from_db['node']
+
+        db.enqueue_delete(node, instance_from_db['uuid'], 'deleted')
+
+        start_time = time.time()
+        while time.time() - start_time < config.parsed.get('API_ASYNC_WAIT'):
```
Comment (mandoonandy) [inline shakenfist/external_api/app.py:580]: This is part of a greater system design issue: Now we are queuing do we need to wait at all?

File: shakenfist/external_api/app.py [hunk 3]
```diff
@@ -876,10 +856,23 @@ def delete(self, confirm=False, namespace=None):
             if instance['state'] == 'deleted':
                 continue
 
-            instance_from_db_virt = virt.from_db(instance['uuid'])
-            _delete_instance(instance['uuid'], instance_from_db_virt)
+            # If this instance is not on a node, just do the DB cleanup locally
```
Comment (mandoonandy) [inline shakenfist/external_api/app.py:859]: Maybe place this in a function to stop code repetition.

File: shakenfist/daemons/main.py [hunk 4]
```diff
@@ -58,7 +59,7 @@ def restore_instances():
                 i.create()
             except Exception as e:
                 util.ignore_exception('restore instance %s' % instance, e)
-                db.update_instance_state(instance, 'error')
+                db.enqueue_delete(node, instance, 'error')
```
Comment (mandoonandy) [inline shakenfist/daemons/main.py:62]: Lint caught this use of node

**Review by mandoonandy — CHANGES_REQUESTED** [review]
Looks good except lint caught one thing and I'm just nit-picking.

File: shakenfist/daemons/main.py [hunk 5]
```diff
@@ -146,20 +147,29 @@ def main():
...
+    LOG.info('queues pid is %d' % queues_pid)
 
     restore_instances()
 
     procnames = {
         external_api_pid: 'external api',
         net_pid: 'network monitor',
         resource_pid: 'resource monitor',
-        cleaner_pid: 'cleaner'
+        cleaner_pid: 'cleaner',
```
[the remark this answers is not in the document]
Comment (mikalstill) [reply shakenfist/daemons/main.py:171]: This gets rewritten later in this series. Can we suspend disbelief until then?

File: shakenfist/daemons/triggers.py (hunk 1 above)
Comment (mikalstill) [reply shakenfist/daemons/triggers.py:68]: Done.

File: shakenfist/external_api/app.py (hunk 2 above)
Comment (mikalstill) [reply shakenfist/external_api/app.py:580]: So this is to maintain the behaviour of the API blocking until the operation completes. I actually re-write this later in the series to replace it with blocking / polling in the client, but that requires that we tweak all the client libraries.

File: shakenfist/external_api/app.py (hunk 3 above)
Comment (mikalstill) [reply shakenfist/external_api/app.py:859]: I'm confused... This code isn't repeated?

File: shakenfist/daemons/main.py (hunk 4 above)
Comment (mikalstill) [reply shakenfist/daemons/main.py]: Fixed.

**Pull request closed without merging**